### ISSUE TITLE
Compiler crash: error: expected: !is_optimizing() || !instr->HasUnmatchedInputRepresentations()

### ISSUE THREAD

**Issue by knaeckeKami** [dart-lang/sdk#51019]
Wanted to use an old library of mine and noticed that it cannot be compiler with any dart version after 2.15.0 (stable).

Repro:

```bash
git clone https://github.com/knaeckeKami/barbecue.git
cd barbecue
dart pub get
dart compile exe test/table_border_test.dart
```

Output:

https://gist.github.com/knaeckeKami/bc202626e87171616e629606d55e9e28 (too long for github)

* Dart SDK Version: tested various, any after 2.15.0 will fail. newest tested one: 3.0.0-75.0.dev (dev) (Thu Dec 29 16:50:26 2022 -0800) on "macos_arm64"
* OS: MacOS, Arm64

**Comment by mraleph** [comment]
@alexmarkov Could you take a look at this? Seems like rogue unboxing being applied to an already unboxed value. Probably we modified the graph _after_ the last select representation was already performed. Maybe load forwarding or canonicalizer broke the graph?

**Comment by alexmarkov** [comment]
`AllocationSinking_Sink` pass propagated load and replaced `v1563` with `v33`:

```
    v33 <- StaticCall:114( _List@0150898.filled<0> v26, v32, v5, result_type = T{_List}) T{_List}
    StoreField(v16 . rowHeights = v33)
...
    v1563 <- LoadField:1048(v16 . rowHeights {final}) T{List<int>?}
    v1573 <- LoadField:1048(v1563 . Array.length {final}) [0, 576460752303423487] T{_Smi}
    v1792 <- UnboxInt64:1048([non-speculative], v1573) [0, 576460752303423487] T{int}
```

=>

```
    v1573 <- LoadField:1048(v33 . Array.length {final}) [0, 576460752303423487] T{_Smi}
    v1792 <- UnboxInt64:1048([non-speculative], v1573) [0, 576460752303423487] T{int}
```

After that, the next `Canonicalize` happened only after `SelectRepresentations_Final` and replaced `v1573` with `v32` which apparently has a different representation:

```
    v1792 <- UnboxInt64:1048([non-speculative], v32 T{_Smi}) [0, 576460752303423487] T{int}
```

This made `UnboxInt64` instruction incorrect and caused crash during code generation.

There are a few things to fix and improve here:
1) `Canonicalize` should not replace a value if it has a different representation and `flow_graph->unmatched_representations_allowed()` is false. It is set to false after `SelectRepresentations_Final`.
2) It could be beneficial to have an additional `Canonicalize` between `AllocationSinking_Sink` and `SelectRepresentations_Final` so it could cleanup the IL before representations are frozen.
3) Canonicalization of boxing and unboxing instructions can check if their input has a desired representation already and eliminate boxing/unboxing if that's the case.

**Comment by a-siva** [comment]
@alexmarkov the list you have above I presume is recommendatations to make the code more robust, should we file issues to track those?

**Comment by alexmarkov** [comment]
@a-siva I'm working on the follow-up CL https://dart-review.googlesource.com/c/sdk/+/279274 which would fix items 2 and 3 from the list above. The CL points to this issue, so there's no need to file separate issue(s) for those improvements.
